compiler/src/org.graalvm.compiler.hotspot.aarch64.test/src/org/graalvm/compiler/hotspot/aarch64/test/UseBarriersForVolatileTest.java

```diff
@@ -0,0 +1,296 @@
+ * unique to each run and makeing the test succeed without executing when the property is not set.
+ * Th e relevant system property is not setin the outer JVM but is set by the command line used to
+ * create the subordinate JVM.
+ */
+public class UseBarriersForVolatileTest extends GraalCompilerTest {
+    /**
+     * Tests compilation requested by the VM.
+     */
+    static class Probe {
+        final String substring;
+        final int expectedOccurrences;
+        int actualOccurrences;
+        String lastMatchingLine;
+
+        Probe(String substring, int expectedOccurrences) {
+            this.substring = substring;
+            this.expectedOccurrences = expectedOccurrences;
+        }
+
+        boolean matches(String line) {
+            if (line.contains(substring)) {
+                actualOccurrences++;
+                lastMatchingLine = line;
+                return true;
+            }
+            return false;
+        }
+
+        String test() {
+            return expectedOccurrences == actualOccurrences ? null : String.format("expected %d, got %d occurrences", expectedOccurrences, actualOccurrences);
+        }
+    }
+
+    @Test
+    public void testVarHandlesInSubJVM() throws IOException, InterruptedException {
+        // the criterion for success is that all 4 tests finish ok
+        Probe successProbe = new Probe("OK (2 tests)", 1);
+        Probe testReadProbe = new Probe("Testing: testReadSnippet", 1);
+        Probe testWriteProbe = new Probe("Testing: testWriteSnippet", 1);
+        List<Probe> probes = Arrays.asList(testReadProbe, testWriteProbe, successProbe);
+        List<String> extraOpts = Arrays.asList("-XX:+UseBarriersForVolatile");
+        // run the tests belonging to inner class Internal in a subordinate test JVM
+        testHelper(probes, extraOpts, Internal.class.getName());
+    }
+
+    private static final boolean VERBOSE = Boolean.getBoolean(UseBarriersForVolatileTest.class.getSimpleName() + ".verbose");
+
+    // the name of the JUnit runner class we expect to find on the original command line
+    private static final String JUNIT_MAIN_CLASS_NAME = "com.oracle.mxtool.junit.MxJUnitWrapper";
+
+    // the property used to inhibit execution of desired tests in the main
+    // JVM and enable execution of desired tests in the subordinate JVM
+    private static final String ENABLE_SUBORDINATE_TESTS_PROPERTY = UseBarriersForVolatileTest.class.getName() + ".enable.subordinate.tests";
+
+    /**
+     * Gets the command line used to start the current tests, including all VM arguments, the test
+     * runner class and its option flags. This can be used to spawn an identical test run for the
+     * secondary test class.
+     */
+    public static List<String> getVMTestCommandLine() {
+        List<String> args = getProcessCommandLine();
+        if (args == null) {
+            throw new InternalError();
```

Comment:
it's always nice to have a message in cases where stack traces fail to be reported correctly.
Maybe simply: `"Could not find the command line for the current process"`